server/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionRetryOnClosedNodeIT.java

```diff
@@ -0,0 +1,232 @@
+                                }
+                            }
+                            sender.sendRequest(connection, action, request, options, handler);
+                        }
+                    };
+                }
+            });
+        }
+    }
+
+    public void testRetryOnStoppedTransportService() throws Exception {
+        internalCluster().startMasterOnlyNodes(2);
+        String primary = internalCluster().startDataOnlyNode();
+        assertAcked(prepareCreate("test")
+            .setSettings(Settings.builder()
+                .put(indexSettings())
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
+            ));
+
+        String replica = internalCluster().startDataOnlyNode();
+        String coordinator = internalCluster().startCoordinatingOnlyNode(Settings.EMPTY);
+        ensureGreen("test");
+
+        TestPlugin primaryTestPlugin = getTestPlugin(primary);
+        // this test only provoked an issue for the primary action, but for completeness, we pick the action randomly
+        primaryTestPlugin.testActionName = TestAction.ACTION_NAME  + (randomBoolean() ? "[p]" : "[r]");
+        logger.info("--> Test action {}, primary {}, replica {}", primaryTestPlugin.testActionName, primary, replica);
+
+        AtomicReference<Object> response = new AtomicReference<>();
+        CountDownLatch responseLatch = new CountDownLatch(1);
+        CountDownLatch doneLatch = new CountDownLatch(1);
+        client(coordinator).execute(TestAction.TYPE, new Request(new ShardId(resolveIndex("test"), 0)),
+            ActionListener.wrap(
+                r -> {
+                    assertTrue(response.compareAndSet(null, r));
+                    responseLatch.countDown();
+                    doneLatch.countDown();
+                },
+                e -> {
+                    assertTrue(response.compareAndSet(null, e));
+                    responseLatch.countDown();
+                    doneLatch.countDown();
+                }));
+
+
+        assertTrue(primaryTestPlugin.actionRunningLatch.await(10, TimeUnit.SECONDS));
+
+        MockTransportService primaryTransportService = (MockTransportService) internalCluster().getInstance(TransportService.class,
+            primary);
+        // we pause node after TransportService has moved to stopped, but before closing connections, since if connections are closed
+        // we would not hit the transport service closed case.
+        primaryTransportService.addOnStopListener(() -> {
+            primaryTestPlugin.actionWaitLatch.countDown();
+            try {
+                assertTrue(doneLatch.await(10, TimeUnit.SECONDS));
+            } catch (InterruptedException e) {
+                throw new AssertionError(e);
+            }
+        });
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(primary));
+
+        assertTrue(responseLatch.await(10, TimeUnit.SECONDS));
```

Comment:
`responseLatch` and `doneLatch` are effectively the same? Maybe just wait for `doneLatch` here as well and get rid of `responseLatch`?

Reply:
Thanks, too many iterations over various interaction patterns, this is clearly redundant, fixed in 4e2cc6dd477ce67f16e3d417807b24aae24b4f42